File: Basic/Python/Day-10/BracketsBalancing.py
class Stack:
    def __init__(self):
        self.items = []

    def push(self, value):
        self.items.append(value)

    def pop(self):
        if not self.items:
            return None
        return self.items.pop()

    def peek(self):
        if not self.items:
            return None
        return self.items[-1]

    def size(self):
        return len(self.items)

    def is_empty(self):
        return not self.items

    def show(self):
        return self.items

def balance_brackets(expr):
    pairs = {
        ')': '(',
        '}': '{',
        ']': '['
    }
    stack = Stack()

    for ch in expr:
        if ch in '({[':
            stack.push(ch)
        elif ch in ')}]':
            if stack.is_empty() or stack.pop() != pairs[ch]:
                return False

    return stack.is_empty()

if __name__ == '__main__':
    print(balance_brackets('({[]})'))
    print(balance_brackets('([)]'))
    print(balance_brackets('(([]))'))

File: Basic/Java/Day-10/BracketsBalancing.java
public class BracketsBalancing {
    public static void main(String[] args) {
        
    }

    public static boolean balance_brackets(String s){
        java.util.Stack<Character> stack = new java.util.Stack<>();
        char[] pairs = {
            ')', '(',
            '}', '{',
            ']', '['
        };
        
        for(char c : s.toCharArray()){
            if(c == '(' || c == '{' || c == '['){
                stack.push(c);
            } else if(c == ')' || c == '}' || c == ']'){
                if(stack.isEmpty()) return false;
                char top = stack.pop();
                if(!isMatching(top, c)) return false;
            }
        }
        
        return stack.isEmpty();
    }
    
    private static boolean isMatching(char open, char close){
        return (open == '(' && close == ')') ||
               (open == '{' && close == '}') ||
               (open == '[' && close == ']');
    }
}
